### PR TITLE
Beam sync

### PR DESCRIPTION
**Pull request by tkstanczak**
fixes the issue where blocks could not be retrieved in the fast synced mode via JSON RPC
improves the eth module JSON RPC consistency and performance across all the requests
specifically improves eth_getBlockBy... by allowing header to be retrieved when no txs are needed
introduces beam sync based sync prototype

@LukaszRozmej since this contains important improvements for JSON RPC - would you be able to review the AuRa tests that are failing and merge when everything is passing?

I want to merge it and keep testing / fixing so we can prepare the 1.5.9 somewhere late Monday / early Tuesday - there will be a need to run a few fast sync - Clique / AuRa / mainnet

### CODE REVIEW

File: src/Nethermind/Nethermind.Blockchain/BlockTree.cs [hunk 1]
```diff
@@ -953,33 +959,31 @@ public void UpdateMainChain(Block[] processedBlocks)
             }
 #endif
 
-            long lastNumber = ascendingOrder ? processedBlocks[processedBlocks.Length - 1].Number : processedBlocks[0].Number;
+            long lastNumber = ascendingOrder ? processedBlocks[^1].Number : processedBlocks[0].Number;
```
Comment (tkstanczak) [inline src/Nethermind/Nethermind.Blockchain/BlockTree.cs]: no change here beside the using change and ^1

File: src/Nethermind/Nethermind.Blockchain/Find/IBlockFinder.cs [hunk 2]
```diff
@@ -30,6 +30,8 @@ public interface IBlockFinder
         
         Keccak PendingHash { get; }
         
+        BlockHeader Head { get; }
```
Comment (tkstanczak) [inline src/Nethermind/Nethermind.Blockchain/Find/IBlockFinder.cs]: speed up for RPC

File: src/Nethermind/Nethermind.Network/PeerManager.cs [hunk 3]
```diff
@@ -493,7 +493,7 @@ private void SelectAndRankCandidates()
             bool hasOnlyStaticNodes = false;
             if (!_currentSelection.PreCandidates.Any() && _staticNodes.Values.Any())
             {
-                _currentSelection.Candidates.AddRange(_staticNodes.Values);
+                _currentSelection.Candidates.AddRange(_staticNodes.Values.Where(sn => !_activePeers.ContainsKey(sn.Node.Id)));
```
Comment (tkstanczak) [inline src/Nethermind/Nethermind.Network/PeerManager.cs]: check if this is ok with you

File: src/Nethermind/Nethermind.AuRa.Test/Reward/AuRaRewardCalculatorTests.cs [hunk 4]
```diff
@@ -112,11 +112,11 @@ public void calculates_rewards_correctly_after_contract_transition(long blockNum
         public void calculates_rewards_correctly_for_ommers(long blockNumber, long expectedReward)
         {
             _block.Header.Number = blockNumber;
-            _block.Body.Ommers = new[]
+            _block.Body = new BlockBody(_block.Body.Transactions, new[]
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.AuRa.Test/Reward/AuRaRewardCalculatorTests.cs]: Why is constructing new BlockBody needed?

File: src/Nethermind/Nethermind.Logging/LoggerExtensions.cs [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+//  You should have received a copy of the GNU Lesser General Public License
+//  along with the Nethermind. If not, see <http://www.gnu.org/licenses/>.
+
+using System;
+
+namespace Nethermind.Logging
+{
+    public static class LoggerExtensions
+    {
+        public static void Error(this ILogger logger, bool overrideCondition, LogLevel originalLevel, string message)
```
[the remark this answers is not in the document]
Comment (tkstanczak) [reply src/Nethermind/Nethermind.Logging/LoggerExtensions.cs]: this is not planned to stay - just for me to quickly detect before the merge

File: src/Nethermind/Nethermind.JsonRpc/Modules/Eth/EthModule.cs [hunk 6]
```diff
@@ -119,123 +119,115 @@ public ResultWrapper<IEnumerable<Address>> eth_accounts()
...
+            Account account = _blockchainBridge.GetAccount(address, header.StateRoot);
+            if (account == null)
             {
-                return result;
+                return ResultWrapper<byte[]>.Success(Bytes.Empty);
             }
 
-            if (_logger.IsTrace) _logger.Trace($"eth_getCode request {address}, {blockParameter}, result: {result.Data.ToHexString(true)}");
-            return result;
+            var code = _blockchainBridge.GetCode(account.CodeHash);
```
[the remark this answers is not in the document]
Comment (tkstanczak) [reply src/Nethermind/Nethermind.JsonRpc/Modules/Eth/EthModule.cs]: account.CodeHash is on a valid root

File: src/Nethermind/Nethermind.AuRa.Test/Reward/AuRaRewardCalculatorTests.cs (hunk 4 above)
Comment (tkstanczak) [reply src/Nethermind/Nethermind.AuRa.Test/Reward/AuRaRewardCalculatorTests.cs]: making it immutable - we use Body.Empty and we  modified static Body.Empty which broke tests

File: src/Nethermind/Nethermind.Runner/Ethereum/Steps/SetupKeyStore.cs [hunk 7]
```diff
@@ -24,7 +24,6 @@
 
 namespace Nethermind.Runner.Ethereum.Steps
 {
-    [RunnerStepDependency]
```
[the remark this answers is not in the document]
Comment (tkstanczak) [reply src/Nethermind/Nethermind.Runner/Ethereum/Steps/SetupKeyStore.cs]: ok

**Pull request merged**